Make sure notebook revision is based on latest before autosaving

wlach opened the pull request:
@hamilton @bcolloran This is not quite ready for review yet, but just wanted to plant my flag in the ground around the autosave code. There is a fair amount of refactoring going on here, and I'd prefer if people didn't touch this area of the codebase (beyond critical bug fixes) until we get this landed.

This mostly implements the approach described here: https://github.com/iodide-project/iodide/issues/1569#issuecomment-470270951

Basically this does a check every time the iodide tab/view is opened. If it detects an inconsistency between the current revision id and the one on the server, a header message is displayed saying that editing is disabled and the user should reload the page.

Still todo:

* More complete tests (which will incidentally test the business logic around saving in general)
* Add some documentation on this error state that we can refer people to

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Documentation**: If this feature has or requires documentation, the relevant docs have been updated.
- [x] **Changelog**: This PR updates the [changelog](../CHANGELOG.md) with any user-visible changes.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not

wlach commented:
Ok, this still needs documentation and (probably) some more tests, but I thought I'd ask for feedback on it if people have time. tl;dr: we now check that the notebook revision we have locally is up-to-date when switching back to the tab/browser, and disable editing and saving if we find out that it isn't. Here's what the looks like:

![image](https://user-images.githubusercontent.com/20569/54160572-4e066a00-4426-11e9-9206-bad93682ab9c.png)

Because it seemed like a natural extension, I also made this change disable autosaving when viewing an older revision: #1593.

There are lots of logical extensions to this PR which I would rather do as followups:

1. Allow user to revert/override saving a revision.
2. Queue a save automatically when switching tabs (this should help make sure that any links shared contain the absolute latest version)
3. ... probably some other small things I can't think of atm ...

bcolloran commented:
hey @wlach working through a detailed review (looking good so far), but had a couple general thoughts that are independent of the specifics in the review--

1) i want to fix all the circular imports in our code (they can cause surprising bugs and races, and circular deps are just a code smell). the other day i was starting to look at fixing the circularities around autosave stuff (which predate this PR). I don't want to collide with your work, so as long as you're working in this area, would you mind taking a pass at resolving those? It will probably just involve moving a few functions to their own files and untangling the import graph. If there is a genuine circularity, you might need to rethink some of the architecture (but if there is a genuine circularity, we _should_ rethink that chunk of the architecture)
```
WARNING in Circular dependency detected:
src/actions/actions.js -> src/tools/local-autosave.js -> src/actions/actions.js

WARNING in Circular dependency detected:
src/tools/local-autosave.js -> src/actions/actions.js -> src/tools/local-autosave.js
```

2) this is a solid next step, but i do want to make sure we're not painting ourselves into an architectural corner; I don't see anything in here that looks like such a corner, but i want your thoughts: longer term, i maintain that the right UX is to allow branching off code from previous points in the revision tree ant any point by visiting an old version and starting to edit. architecturally, i _think_ the way to do that is to have every version point to a previous parent version. obviously moving to the branchable approach will require making some changes and overhauling some stuff in here, but do you see anything that will lock us into a decision that would make that harder to do down the road?

thanks will!

bcolloran commented:
sounds good re (1) and (2), thanks will!

hope to get the rest of the review completed soon

hamilton commented on src/components/menu/header-messages.jsx:
```diff
@@ -87,6 +88,18 @@ export class HeaderMessagesUnconnected extends React.Component {
+      case "NOTEBOOK_REVISION_ID_OUT_OF_DATE":
+        content = (
+          <React.Fragment>
+            You are viewing an old version of this notebook, editing is
```
oh nvm I think your langauge is actually better but would remove the comma

wlach replied:
Seems a bit like a run on sentence to me that way. How about `You are viewing an old version of this notebook: editing is disabled.` ?

hamilton replied:
I mean replace with a period

wlach commented on src/index.jsx:
```diff
@@ -49,8 +50,24 @@ window.addEventListener("message", listenForEvalFramePortReady, false);
+
+document.addEventListener(
+  "visibilitychange",
+  () => {
```
It shouldn't be a problem, but it doesn't sound like great behavior. I'll see if we can avoid this somehow.

wlach replied:
Hmm, I can't reproduce this.

hamilton replied:
hmm, can't seem to reproduce this either. Strange. Well, it works out either way!

wlach commented on src/actions/actions.js:
```diff
@@ -385,6 +386,40 @@ export function saveNotebookToServer(appMsg = true) {
+      // out of date
+      return;
+    }
+    dispatch({ type: "CHECKING_NOTEBOOK_REVISION_IS_LATEST", state: true });
```
I think I was just following the example I saw elsewhere. Sure, I can change this.

wlach commented:
@hamilton yeah, this doesn't completely solve all possible problems or scenarios. 

As I mentioned a logical followup to this is to autosave the document when switching contexts. I also have a half-written patch that checks what revision an autosaved one is based on, and will refuse to save if that revision is not the latest one.

I'd prefer to do those things in followups, rather than expand the scope of this PR further. I think it will be easier to get things right this way. My main concern is making sure the functionality here works as expected. :)

wlach commented on src/tools/server-autosave.js:
```diff
@@ -3,43 +3,58 @@ import {
+      const { username: notebookOwner } = state.notebookInfo;
+      const { name: thisUser } = state.userData;
+      let validAutosave = false;
+      console.log("Saving notebook");
```
Yup, that + other console.logs were just for debugging purposes.

wlach commented on src/tools/server-autosave.js:
```diff
@@ -3,43 +3,58 @@ import {
+      if (
+        notebookChangedSinceSave(state) && // has the notebook changed?
+        thisUser !== undefined && // is this a logged-in-user?
+        notebookOwner === thisUser // is this notebook owned by the current user?
```
Yeah, I think I agree. Would be much easier to properly test this way. I think I'd prefer to do that in a followup if that's ok though.

bcolloran replied:
as a convention: let's try to always remember to put in FIXME comments in cases like this where we know there's a little thing that we want to circle back on, but we don't want to get it during the PR in question. i've found it hand to be able to search the code for fixmes and bang out quick PRs that tackle a bunch of these things

wlach replied:
I added a FIXME to this PR.

bcolloran commented on src/state-schemas/state-schema.js:
```diff
@@ -229,11 +229,17 @@ export const stateProperties = {
       }
     }
   },
+  checkingRevisionIsLatest: {
```
via ctrl-F, I'm spotting where this piece of state is set, but not where it is actually used. Maybe this is left over from an earlier idea?

wlach replied:
good catch, I think there's actually some bugs here (or at least poorly specified behaviour). looking into it...

bcolloran replied:
ok np. i'll hold off on more review until you give me the go ahead. thanks @wlach

wlach commented on src/tools/autosave.js:
```diff
@@ -1,123 +1,65 @@
+} from "./server-autosave";
 import { connectionModeIsServer } from "./server-tools";
-import { setPreviousAutosave, loadAutosave } from "../actions/actions";
+import { store } from "../store";
```
In general I agree, unfortunately in this case we kind of need to do this, as the redux store might be out of date by the time the callback fires. Same with the case you noticed below. You'll note that the main function `updateAutoSave` in this file takes a state parameter, rather than using the store.

That said, it's possible there is a better way of organizing this than what I came up with. Suggestions welcome.

wlach replied:
Actually, looking at this again, I don't think we need to import store. Will fix.

wlach commented:
@hamilton @bcolloran this is ready for another look. I'm not entirely happy with the spaghetti type mess that's in `src/tools/autosave.js` and friends, though at least there is now are tests for the most complicated/important part (the part that checks whether we should save or not). If you have suggestions on how to make things more straightforward, I would welcome them.

wlach commented:
@robhudson hey, would you mind taking a quick look at the python/django/drf parts of this PR? thanks :)

robhudson commented on server/notebooks/views.py:
```diff
@@ -31,13 +31,16 @@ def notebook_view(request, pk):
     notebook = get_object_or_404(Notebook, pk=pk)
     if "revision" in request.GET:
         revision = get_object_or_404(NotebookRevision, pk=int(request.GET["revision"]))
```
I don't know if there's any fear but if this can't be turned into an `int` this could 500 the API.

robhudson replied:
Also, do you want to make sure this is revision is assigned to the correct `Notebook` also? I.e. `get_object_or_404(NotebookRevision, notebook=notebook, pk=int(request.GET["revision"]))`

wlach replied:
both of these are good points, will fix before landing.

robhudson commented on server/notebooks/views.py:
```diff
@@ -31,13 +31,16 @@ def notebook_view(request, pk):
     notebook = get_object_or_404(Notebook, pk=pk)
     if "revision" in request.GET:
         revision = get_object_or_404(NotebookRevision, pk=int(request.GET["revision"]))
+        latest_revision_id = notebook.revisions.values_list("id", flat=True).first()
```
I didn't check but I'm assuming this has an ordering on the model class so the first `id` is the most recent? There's also `notebook.revisions.get_latest_by("created").id` which explicitly provides the date field.

wlach replied:
that's correct on the ordering front :) i think your solution makes this more explicit, so i'll switch to that approach

hamilton commented on src/index.jsx:
```diff
@@ -49,8 +50,23 @@ window.addEventListener("message", listenForEvalFramePortReady, false);
-checkForServerAutosave(store);
+
+// re-apply any locally saved changes if applicable
+checkForLocalAutoSave(store);
```
Reading this bit of code, I wonder if something like `applyLocallySavedChanges(store)` might be a better name for what this is doing. That may pigeonhole this fcn for later use, though - I leave it to you.

wlach replied:
Yeah, this sounds right. I'll rename the function before landing.

wlach commented:
@bcolloran there are several core ideas in this PR:

1) Track the server revision this notebook is based on
2) Only attempt to autosave anything at all if the notebook is up to date with what is on the server (i.e. the revision on the server matches the one the client is based on)
3) Make server autosave *dependent* on a local autosave succeeding, rather than having them be two independent operations

This [mermaid chart](https://mermaidjs.github.io/mermaid-live-editor/#/view/eyJjb2RlIjoiZ3JhcGggVERcbkFbSGFzIHRoZXJlIGJlZW4gYW4gaW50ZXJhY3Rpb24gd2l0aCB0aGUgcmVkdXggc3RvcmVdXG5BIC0tPnxOb3wgQltEbyBub3RoaW5nXVxuQSAtLT58WWVzfCBDW1NjaGVkdWxlIGEgbG9jYWwgYXV0b3NhdmUgb3BlcmF0aW9uIGluIG9uZSBzZWNvbmRdXG5DIC0tPiB8T25lIHNlY29uZCBlbGFwc2VzfCBEW0FyZSB3ZSBjaGVja2luZyBub3RlYm9vayBjb25zaXN0ZW5jeV1cbkQgLS0-IHxZZXN8IEVbRG8gbm90aGluZ11cbkQgLS0-IHxOb3wgRltBcmUgd2UgYmFzZWQgb24gdGhlIGxhdGVzdCB2ZXJzaW9uIG9uIHRoZSBzZXJ2ZXIgYXMgZmFyIGFzIHdlIGtub3ddXG5GIC0tPiB8Tm98IEdbRG8gbm90aGluZ11cbkYgLS0-IHxZZXN8IEhbSGFzIHRoZXJlIGJlZW4gYSBjaGFuZ2UgdG8gdGhlIGxvY2FsIGpzbWRdXG5IIC0tPiB8Tm98IElbRG8gbm90aGluZ11cbkggLS0-IHxZZXN8IEpbU2F2ZSBqc21kIGluIGxvY2FsIGluZGV4ZGIsIHNjaGVkdWxlIHNlcnZlciBhdXRvc2F2ZSBpbiAzMCBzZWNvbmRzXVxuSiAtLT4gfDMwIHNlY29uZHMgZWxhcHNlfCBLW0FyZSB3ZSBsb2dnZWQgYW5kIGRvIHdlIG93biB0aGUgbm90ZWJvb2s_XVxuSyAtLT4gfE5vfCBMW0RvIG5vdGhpbmddXG5LIC0tPiB8WWVzfCBNW1N1Ym1pdCBzYXZlIHRvIHNlcnZlcl1cbk0gLS0-IE5bRGlkIHdlIHN1Y2NlZWQ_XVxuTiAtLT4gfFllc3wgT1tZYXkhIFVwZGF0ZSBvdXIgaW50ZXJuYWwgYWNjb3VudGluZyBvZiB0aGUgcmV2aXNpb24gd2UncmUgYmFzZWQgb25dXG5OIC0tPiB8Tm98IFBbT2ggbm8uIEZvciBub3cgYXNzdW1lIHRoaXMgd2FzIGEgdHJhbnNpZW50IGVycm9yIGFuZCByZXRyeV1cbiIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In19) tries to encompass the logic of this PR. If you have any questions let me know-- I'd prefer to deal with all review comments in one fell swoop if possible, rather than do another back & forth.

bcolloran commented:
thanks @wlach that is super helpful!

wlach commented:
Note: I in fact have not added any documentation to this PR-- I'm honestly not sure if it's necessary-- I think the exception case here (notebook is determined to be out of date after switching back to a tab/browser after a while or when viewing an old revision) should be rare and I think the error message is self-explanatory.

bcolloran commented:
that flow chart is super helpful @wlach. looking at it helps me see a few things: 

1) we should not be subscribing to the store at all -- rather than looking at every single change to the store and doing stuff with saves based on that, we should trigger the autosave flow from with the specific redux actions that should trigger a save attempt -- at this point, i think that is only `changePageTitle` and `updateJsmdContent`, b/c the title and the jsmd are the only things we actually save. this will also make the step "Has there been a change to the local jsmd" irrelevant and we can factor out the `notebookChangedSinceSave` code, because we'll only be firing this action on those changes.

2) let's use a debouncing function to handle the debouncing of these save actions rather than writing custom functions in both cases. this should clean up the code immensely. there is no reason you should have to write a setTimeout by hand in any of the functions that are actually doing work here. here are a few things to look at:
https://github.com/jshanson7/debounce-action  (this one seems fine)
https://github.com/RemeJuan/debounce-redux-dispatch
https://gist.github.com/krstffr/245fe83885b597aabaf06348220c2fe9

2) basically that whole flow could be encapsulated cleanly in one async function, because all of these things are conceptually ordered and there are no complex branches or cycles:
```
runAutosaveFlow() {
return async (dispatch, getState()) => {
  dispatch(debouncedLocalAutosave())
  isLatest = await dispatch(isLocalNotebookOnLatestServerVersion())
  if (isLatest && loggedInAndOwnThisNB(getStore())){
    dispatch(debouncedServerAutosave())
  }
}
}
```
i'm 93% sure we can totally factor out things like the extra footwork we have to do with checkNotebookConsistency if we can consolidate the whole set of operations in one place.

3) this will involve converting some of your "helper functions" that are getting `store` as an explicit argument to redux thunk actions, but that should not be a big deal

i know this is a moderate sized refactor, but i'm quite sure that by inverting this we can end up with much cleaner and more maintainable code and untangle some of the spaghetti-feeling bits.

unfortuntately, i can't promise that there will not be another round of back and forth. i believe we can clean this up substantially, but since i'm asking for non-trivial design changes, another round of review will be needed

wlach commented:
Sure, this sounds great if we can make it work. There's a bit of a balance here between trying to do everything in one go vs. getting incremental improvements (I'd argue that the existing code is a structural improvement over what's in there currently), but I can try some of the above suggestions and see how it goes.

P.S. The flow in (3) seems not quite correct to me (we shouldn't try to locally autosave at all if we don't have the latest revision)

bcolloran commented:
sure sure, that snippet of psuedo-code was just a "for instance" -- details will need to be worked out to get the correct behavior. point being, though: based on the diagram you drew, it's really a very linear flow where each item depends on the next, and so conceptually that maps very nicely to a sequence of `await` statements rather than a tree of functions that call each other to a moderate nesting depth.

I'm inclined to try to iterate on this PR a bit more rather than deferring and building on shaky foundations until the next time we circle back around to it, because i think it's not actually too much more to invert things as described, and having spent a bunch of time this weekend going through code that had gotten to be a big cruft pile, i think we can save future headaches with a moderate incremental investment right now :-)

wlach commented:
@bcolloran could you have another look? 

I removed the store hook stuff and moved the main autosave function into actions (got to clean that up... but not here) and converted it into a debounced redux thunk, which is called by the two actions you mentioned actually modified the jsmd.

I also fixed some minor things @robhudson noticed earlier on the server side.

I think this is a good point to commit work before moving further with refactoring / new features.

wlach commented on src/tools/server-autosave.js:
```diff
@@ -1,45 +1,46 @@
-  store.dispatch(setMostRecentSavedContent());
-  setInterval(async () => {
-    const state = store.getState();
+const updateServerAutosave = _.throttle(
```
Yeah, it wasn't a redux thunk so I stopped short of that. However, on reflection it was pretty trivial to make it one, so I did so and moved the file over to a file called `server-actions.js` (which I figure could incorporate some other actions which are server-specific).

wlach commented on src/reducers/notebook-reducer.js:
```diff
@@ -65,9 +65,9 @@ const notebookReducer = (state = newNotebook(), action) => {
-    case "SET_PREVIOUS_AUTOSAVE": {
-      const { hasPreviousAutoSave } = action;
-      return Object.assign({}, state, { hasPreviousAutoSave });
+    case "SET_PREVIOUS_Autosave": {
```
🤦‍♂️ oops

wlach commented:
@bcolloran ok, should be good to go now (I hope)

bcolloran commented:
thanks will! i shall merge! :-)
